File: axl/validation.py
import re

from .ir import (
    Agent,
    Annotation,
    Binary,
    Emit,
    Expression,
    Forget,
    Function,
    FunctionCall,
    If,
    Instruction,
    Let,
    ListExpression,
    Literal,
    MapExpression,
    MemoryWrite,
    Program,
    Recall,
    Return,
    Run,
    ToolCall,
    UiNode,
    UiView,
    Variable,
    While,
    Workflow,
)
from .type_names import validate_type_name
from .ui_registry import ANNOTATION_KINDS, COMPONENTS


class ValidationError(ValueError):
    pass


INSTRUCTION_TYPES = (
    MemoryWrite,
    Forget,
    Let,
    Return,
    Emit,
    If,
    While,
    Agent,
    Workflow,
    Run,
    Function,
    Annotation,
    UiView,
)
EXPRESSION_TYPES = (
    Literal,
    Variable,
    Recall,
    ToolCall,
    FunctionCall,
    ListExpression,
    MapExpression,
    Binary,
)
OPERATORS = {"+", "-", "*", "/", "==", "!=", ">", "<", ">=", "<="}
MAX_NESTING_DEPTH = 256
MAX_CALL_DEPTH = 256
RESERVED = {
    "agent",
    "call",
    "else",
    "emit",
    "end",
    "false",
    "forget",
    "fn",
    "if",
    "import",
    "let",
    "memory",
    "meta",
    "recall",
    "return",
    "run",
    "true",
    "uses",
    "while",
    "workflow",
}


def _identifier(value, label: str = "identifier") -> None:
    if not isinstance(value, str) or not re.fullmatch(r"[A-Za-z_][A-Za-z0-9_]*", value):
        raise ValidationError(f"invalid {label} identifier")
    if value in RESERVED:
        raise ValidationError(f"reserved {label} identifier '{value}'")


def _qualified_identifier(value, label: str) -> None:
    if not isinstance(value, str):
        raise ValidationError(f"invalid {label} identifier")
    for part in value.split("."):
        _identifier(part, label)


def _type_name(value) -> None:
    try:
        validate_type_name(value)
    except (TypeError, ValueError) as error:
        raise ValidationError(str(error)) from error


def _is_value(value) -> bool:
    return type(value) in (str, int, bool)


def validate(program: Program) -> None:
    if not isinstance(program, Program):
        raise ValidationError("IR root must be Program")
    if not isinstance(program.instructions, tuple):
        raise ValidationError("program instructions must be an array")
    _validate_nesting(program.instructions)

    names: set[str] = set()
    view_ids: set[int] = set()
    runnables: dict[str, Agent | Workflow] = {}
    for instruction in program.instructions:
        _require_instruction(instruction)
        if isinstance(instruction, (Agent, Workflow)):
            _identifier(instruction.name, "runnable")
            if instruction.name in names:
                raise ValidationError(f"duplicate runnable '{instruction.name}'")
            names.add(instruction.name)
            runnables[instruction.name] = instruction
        elif isinstance(instruction, Function):
            _qualified_identifier(instruction.name, "function")
        elif isinstance(instruction, UiView):
            if type(instruction.view_id) is not int or instruction.view_id < 1:
                raise ValidationError("UI view id must be a positive integer")
            if instruction.view_id in view_ids:
                raise ValidationError(f"duplicate UI view id '{instruction.view_id}'")
            view_ids.add(instruction.view_id)

    for instruction in program.instructions:
        _validate_instruction(instruction, names, top_level=True)
    _validate_call_graph(runnables)


def _validate_nesting(instructions) -> None:
    stack = [(instruction, 1) for instruction in instructions]
    while stack:
        node, depth = stack.pop()
        if depth > MAX_NESTING_DEPTH:
            raise ValidationError(f"program nesting depth exceeds {MAX_NESTING_DEPTH}")
        if isinstance(node, (Agent, Workflow, Function)):
            stack.extend((child, depth + 1) for child in node.body)
        elif isinstance(node, While):
            stack.append((node.condition, depth + 1))
            stack.extend((child, depth + 1) for child in node.body)
        elif isinstance(node, If):
            stack.append((node.condition, depth + 1))
            stack.extend((child, depth + 1) for child in node.body)
            stack.extend((child, depth + 1) for child in node.else_body)
        elif isinstance(node, (MemoryWrite, Let, Return, Emit)):
            stack.append((node.value, depth + 1))
        elif isinstance(node, Binary):
            stack.append((node.left, depth + 1))
            stack.append((node.right, depth + 1))
        elif isinstance(node, (ToolCall, FunctionCall)):
            stack.extend((argument, depth + 1) for argument in node.arguments)
        elif isinstance(node, ListExpression):
            stack.extend((item, depth + 1) for item in node.items)
        elif isinstance(node, MapExpression):
            stack.extend((item, depth + 1) for entry in node.entries for item in entry)
        elif isinstance(node, UiView):
            stack.append((node.root, depth + 1))
        elif isinstance(node, UiNode):
            stack.extend((child, depth + 1) for child in node.children)
            stack.extend((item.value, depth + 1) for item in node.properties)


def _require_instruction(instruction) -> None:
    if not isinstance(instruction, INSTRUCTION_TYPES):
        raise ValidationError(
            f"invalid instruction node '{type(instruction).__name__}'"
        )


def _validate_block(body, names: set[str]) -> None:
    if not isinstance(body, tuple):
        raise ValidationError("instruction body must be an array")
    for child in body:
        _require_instruction(child)
        _validate_instruction(child, names, top_level=False)


def _validate_instruction(
    instruction: Instruction, names: set[str], *, top_level: bool
) -> None:
    if isinstance(instruction, (Agent, Workflow, Function)):
        if not top_level:
            raise ValidationError("agent and workflow declarations must be top-level")
        if isinstance(instruction, Agent):
            if not isinstance(instruction.tools, tuple):
                raise ValidationError("agent tools must be an array of names")
            for tool in instruction.tools:
                _identifier(tool, "tool")
            if len(instruction.tools) != len(set(instruction.tools)):
                raise ValidationError(
                    f"duplicate tool grant in agent '{instruction.name}'"
                )
        if isinstance(instruction, Function):
            if not isinstance(instruction.parameters, tuple):
                raise ValidationError("function parameters must be an array")
            _type_name(instruction.return_type)
            for parameter in instruction.parameters:
                _identifier(parameter.name, "parameter")
                _type_name(parameter.type_name)
        _validate_block(instruction.body, names)
    elif isinstance(instruction, Run):
        _identifier(instruction.name, "runnable")
        if instruction.name not in names:
            raise ValidationError(f"unknown runnable '{instruction.name}'")
    elif isinstance(instruction, If):
        _validate_expression(instruction.condition)
        _validate_block(instruction.body, names)
        _validate_block(instruction.else_body, names)
    elif isinstance(instruction, While):
        _validate_expression(instruction.condition)
        _validate_block(instruction.body, names)
    elif isinstance(instruction, MemoryWrite):
        _identifier(instruction.key, "memory")
        _identifier(instruction.source, "source")
        _validate_expression(instruction.value)
        if (
            type(instruction.confidence) is not int
            or not 0 <= instruction.confidence <= 100
        ):
            raise ValidationError("memory confidence must be an integer from 0 to 100")
        if instruction.ttl_seconds is not None and (
            type(instruction.ttl_seconds) is not int or instruction.ttl_seconds < 1
        ):
            raise ValidationError("memory ttl must be a positive integer")
    elif isinstance(instruction, Let):
        _identifier(instruction.target, "variable")
        if instruction.type_name is not None:
            _type_name(instruction.type_name)
        _validate_expression(instruction.value)
    elif isinstance(instruction, (Return, Emit)):
        _validate_expression(instruction.value)
    elif isinstance(instruction, Forget):
        _identifier(instruction.key, "memory")
    elif isinstance(instruction, Annotation):
        if not top_level:
            raise ValidationError("annotations must be top-level")
        if instruction.kind not in ANNOTATION_KINDS:
            raise ValidationError(f"unknown annotation kind '{instruction.kind}'")
        if type(instruction.target) is not int or instruction.target < 1:
            raise ValidationError("annotation target must be a positive integer")
        if not isinstance(instruction.value, str) or not instruction.value:
            raise ValidationError("annotation value must be a non-empty string")
    elif isinstance(instruction, UiView):
        if not top_level:
            raise ValidationError("UI views must be top-level")
        _validate_ui_tree(instruction.root)


def _validate_ui_tree(root: UiNode) -> None:
    node_ids: set[int] = set()
    stack = [root]
    while stack:
        node = stack.pop()
        if not isinstance(node, UiNode):
            raise ValidationError("invalid UI node")
        if type(node.node_id) is not int or node.node_id < 1:
            raise ValidationError("UI node id must be a positive integer")
        if node.node_id in node_ids:
            raise ValidationError(f"duplicate UI node id '{node.node_id}'")
        node_ids.add(node.node_id)
        contract = COMPONENTS.get(node.component_id)
        if contract is None:
            raise ValidationError(f"unknown UI component '{node.component_id}'")
        if not all(isinstance(item, tuple) for item in (node.properties, node.events, node.children)):
            raise ValidationError("UI node collections must be arrays")
        property_ids = [item.property_id for item in node.properties]
        if len(property_ids) != len(set(property_ids)):
            raise ValidationError(f"duplicate UI property on node '{node.node_id}'")
        for item in node.properties:
            expected = contract.properties.get(item.property_id)
            if expected is None:
                raise ValidationError(
                    f"property '{item.property_id}' is not valid for component '{node.component_id}'"
                )
            if not isinstance(item.value, Literal):
                raise ValidationError("experimental UI properties must be literal")
            _validate_expression(item.value)
            actual = {str: "string", int: "int", bool: "bool"}[type(item.value.value)]
            if actual != expected:
                raise ValidationError(
                    f"property '{item.property_id}' requires {expected}, got {actual}"
                )
        event_ids = [item.event_id for item in node.events]
        if len(event_ids) != len(set(event_ids)):
            raise ValidationError(f"duplicate UI event on node '{node.node_id}'")
        for item in node.events:
            if item.event_id not in contract.events:
                raise ValidationError(
                    f"event '{item.event_id}' is not valid for component '{node.component_id}'"
                )
            if type(item.action_id) is not int or item.action_id < 1:
                raise ValidationError("UI action id must be a positive integer")
        if node.children and not contract.children:
            raise ValidationError(f"component '{node.component_id}' cannot have children")
        stack.extend(node.children)


def _validate_expression(expression: Expression) -> None:
    if not isinstance(expression, EXPRESSION_TYPES):
        raise ValidationError(f"invalid expression node '{type(expression).__name__}'")
    if isinstance(expression, Literal):
        if not _is_value(expression.value):
            raise ValidationError("literal must be string, integer, or boolean")
        if type(expression.value) is str:
            try:
                expression.value.encode("utf-8")
            except UnicodeEncodeError as error:
                raise ValidationError("invalid Unicode string") from error
    elif isinstance(expression, Binary):
        if expression.operator not in OPERATORS:
            raise ValidationError(f"unknown operator '{expression.operator}'")
        _validate_expression(expression.left)
        _validate_expression(expression.right)
    elif isinstance(expression, ToolCall):
        _identifier(expression.name, "tool")
        if not isinstance(expression.arguments, tuple):
            raise ValidationError("tool arguments must be an array")
        for argument in expression.arguments:
            _validate_expression(argument)
    elif isinstance(expression, FunctionCall):
        _qualified_identifier(expression.name, "function")
        if not isinstance(expression.arguments, tuple):
            raise ValidationError("function arguments must be an array")
        for argument in expression.arguments:
            _validate_expression(argument)
    elif isinstance(expression, ListExpression):
        if not isinstance(expression.items, tuple):
            raise ValidationError("list items must be an array")
        for item in expression.items:
            _validate_expression(item)
    elif isinstance(expression, MapExpression):
        if not isinstance(expression.entries, tuple):
            raise ValidationError("map entries must be an array")
        for entry in expression.entries:
            if not isinstance(entry, tuple) or len(entry) != 2:
                raise ValidationError("map entry must contain key and value")
            _validate_expression(entry[0])
            _validate_expression(entry[1])
    elif isinstance(expression, Variable):
        _identifier(expression.name, "variable")
    elif isinstance(expression, Recall):
        _identifier(expression.key, "memory")


def _validate_call_graph(runnables: dict[str, Agent | Workflow]) -> None:
    graph = {name: _runs(runnable.body) for name, runnable in runnables.items()}
    state: dict[str, int] = {name: 0 for name in graph}

    for root, dependencies in graph.items():
        if state[root] != 0:
            continue
        state[root] = 1
        stack: list[tuple[str, object]] = [(root, iter(dependencies))]
        while stack:
            name, dependencies = stack[-1]
            try:
                dependency = next(dependencies)
            except StopIteration:
                state[name] = 2
                stack.pop()
                continue
            if state[dependency] == 1:
                raise ValidationError(f"workflow cycle detected at '{dependency}'")
            if state[dependency] == 0:
                if len(stack) >= MAX_CALL_DEPTH:
                    raise ValidationError(
                        f"workflow call depth exceeds {MAX_CALL_DEPTH}"
                    )
                state[dependency] = 1
                stack.append((dependency, iter(graph[dependency])))


def _runs(body) -> set[str]:
    result: set[str] = set()
    for instruction in body:
        if isinstance(instruction, Run):
            result.add(instruction.name)
        elif isinstance(instruction, If):
            result |= _runs(instruction.body) | _runs(instruction.else_body)
        elif isinstance(instruction, While):
            result |= _runs(instruction.body)
    return result

File: runtime/axl-core-rs/src/validation.rs
use std::collections::{HashMap, HashSet};

use crate::ir::*;
use crate::type_names;
use crate::ui_registry;

const MAX_NESTING_DEPTH: usize = 256;
const MAX_CALL_DEPTH: usize = 256;

const RESERVED: &[&str] = &[
    "agent", "call", "else", "emit", "end", "false", "forget", "fn", "if",
    "import", "let", "memory", "meta", "recall", "return", "run", "true",
    "uses", "while", "workflow",
];

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ValidationError(pub String);

impl std::fmt::Display for ValidationError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(&self.0)
    }
}

impl std::error::Error for ValidationError {}

fn check_identifier(value: &str, label: &str) -> Result<(), ValidationError> {
    if value.is_empty() || !value.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'_')
        || !value.as_bytes()[0].is_ascii_alphabetic() && value.as_bytes()[0] != b'_'
    {
        return Err(ValidationError(format!("invalid {label} identifier")));
    }
    if RESERVED.contains(&value) {
        return Err(ValidationError(format!("reserved {label} identifier '{value}'")));
    }
    Ok(())
}

fn check_qualified_identifier(value: &str, label: &str) -> Result<(), ValidationError> {
    for part in value.split('.') {
        check_identifier(part, label)?;
    }
    Ok(())
}

fn check_type_name(value: &str) -> Result<(), ValidationError> {
    type_names::validate_type_name(value).map_err(|e| ValidationError(e.to_string()))
}

pub fn validate(program: &Program) -> Result<(), ValidationError> {
    validate_nesting(program)?;
    let mut names = HashSet::new();
    let mut view_ids = HashSet::new();
    let mut runnables: HashMap<String, usize> = HashMap::new();

    for (i, instruction) in program.instructions.iter().enumerate() {
        match instruction {
            Instruction::Agent(agent) => {
                check_identifier(&agent.name, "runnable")?;
                if !names.insert(agent.name.clone()) {
                    return Err(ValidationError(format!("duplicate runnable '{}'", agent.name)));
                }
                runnables.insert(agent.name.clone(), i);
            }
            Instruction::Workflow(wf) => {
                check_identifier(&wf.name, "runnable")?;
                if !names.insert(wf.name.clone()) {
                    return Err(ValidationError(format!("duplicate runnable '{}'", wf.name)));
                }
                runnables.insert(wf.name.clone(), i);
            }
            Instruction::Function(func) => {
                check_qualified_identifier(&func.name, "function")?;
            }
            Instruction::UiView(view) => {
                if view.view_id < 1 {
                    return Err(ValidationError("UI view id must be a positive integer".into()));
                }
                if !view_ids.insert(view.view_id) {
                    return Err(ValidationError(format!("duplicate UI view id '{}'", view.view_id)));
                }
            }
            _ => {}
        }
    }

    for instruction in &program.instructions {
        validate_instruction(instruction, &names, true)?;
    }
    validate_call_graph(program, &runnables)?;
    Ok(())
}

fn validate_nesting(program: &Program) -> Result<(), ValidationError> {
    let mut stack: Vec<(&Instruction, usize)> = program.instructions.iter().map(|i| (i, 1)).collect();
    while let Some((node, depth)) = stack.pop() {
        if depth > MAX_NESTING_DEPTH {
            return Err(ValidationError(format!("program nesting depth exceeds {MAX_NESTING_DEPTH}")));
        }
        match node {
            Instruction::Agent(a) => stack.extend(a.body.iter().map(|i| (i, depth + 1))),
            Instruction::Workflow(w) => stack.extend(w.body.iter().map(|i| (i, depth + 1))),
            Instruction::Function(f) => stack.extend(f.body.iter().map(|i| (i, depth + 1))),
            Instruction::While(w) => {
                stack.extend(w.body.iter().map(|i| (i, depth + 1)));
            }
            Instruction::If(if_inst) => {
                stack.extend(if_inst.body.iter().map(|i| (i, depth + 1)));
                stack.extend(if_inst.else_body.iter().map(|i| (i, depth + 1)));
            }
            Instruction::UiView(v) => {
                // validate UI tree nesting
                let mut ui_stack = vec![(&v.root, depth + 1)];
                while let Some((node, d)) = ui_stack.pop() {
                    if d > MAX_NESTING_DEPTH {
                        return Err(ValidationError(format!("program nesting depth exceeds {MAX_NESTING_DEPTH}")));
                    }
                    ui_stack.extend(node.children.iter().map(|c| (c, d + 1)));
                }
            }
            _ => {}
        }
    }
    Ok(())
}

fn validate_instruction(
    instruction: &Instruction, names: &HashSet<String>, top_level: bool,
) -> Result<(), ValidationError> {
    match instruction {
        Instruction::Agent(agent) => {
            if !top_level {
                return Err(ValidationError("agent and workflow declarations must be top-level".into()));
            }
            for tool in &agent.tools {
                check_identifier(tool, "tool")?;
            }
            if agent.tools.len() != agent.tools.iter().collect::<HashSet<_>>().len() {
                return Err(ValidationError(format!("duplicate tool grant in agent '{}'", agent.name)));
            }
            validate_block(&agent.body, names)?;
        }
        Instruction::Workflow(wf) => {
            if !top_level {
                return Err(ValidationError("agent and workflow declarations must be top-level".into()));
            }
            validate_block(&wf.body, names)?;
        }
        Instruction::Function(func) => {
            if !top_level {
                return Err(ValidationError("agent and workflow declarations must be top-level".into()));
            }
            check_type_name(&func.return_type)?;
            for param in &func.parameters {
                check_identifier(&param.name, "parameter")?;
                check_type_name(&param.type_name)?;
            }
            validate_block(&func.body, names)?;
        }
        Instruction::Run(name) => {
            check_identifier(name, "runnable")?;
            if !names.contains(name) {
                return Err(ValidationError(format!("unknown runnable '{name}'")));
            }
        }
        Instruction::If(if_inst) => {
            validate_expression(&if_inst.condition)?;
            validate_block(&if_inst.body, names)?;
            validate_block(&if_inst.else_body, names)?;
        }
        Instruction::While(w) => {
            validate_expression(&w.condition)?;
            validate_block(&w.body, names)?;
        }
        Instruction::MemoryWrite(mw) => {
            check_identifier(&mw.key, "memory")?;
            check_identifier(&mw.source, "source")?;
            validate_expression(&mw.value)?;
            if mw.confidence < 0 || mw.confidence > 100 {
                return Err(ValidationError("memory confidence must be an integer from 0 to 100".into()));
            }
            if let Some(ttl) = mw.ttl_seconds {
                if ttl < 1 {
                    return Err(ValidationError("memory ttl must be a positive integer".into()));
                }
            }
        }
        Instruction::Let { target, value, type_name } => {
            check_identifier(target, "variable")?;
            if let Some(tn) = type_name { check_type_name(tn)?; }
            validate_expression(value)?;
        }
        Instruction::Return(expr) | Instruction::Emit(expr) => validate_expression(expr)?,
        Instruction::Forget(key) => check_identifier(key, "memory")?,
        Instruction::Annotation(ann) => {
            if !top_level {
                return Err(ValidationError("annotations must be top-level".into()));
            }
            if !ui_registry::annotation_kind_valid(ann.kind) {
                return Err(ValidationError(format!("unknown annotation kind '{}'", ann.kind)));
            }
            if ann.target < 1 {
                return Err(ValidationError("annotation target must be a positive integer".into()));
            }
            if ann.value.is_empty() {
                return Err(ValidationError("annotation value must be a non-empty string".into()));
            }
        }
        Instruction::UiView(view) => {
            if !top_level {
                return Err(ValidationError("UI views must be top-level".into()));
            }
            validate_ui_tree(&view.root)?;
        }
        _ => {}
    }
    Ok(())
}

fn validate_block(body: &[Instruction], names: &HashSet<String>) -> Result<(), ValidationError> {
    for child in body {
        validate_instruction(child, names, false)?;
    }
    Ok(())
}

fn validate_expression(expr: &Expression) -> Result<(), ValidationError> {
    match expr {
        Expression::Literal(Value::String(s)) => {
            // validate UTF-8
            let _ = s.as_bytes();
        }
        Expression::Literal(_) => {}
        Expression::Binary { left, operator, right } => {
            const OPERATORS: &[&str] = &["+", "-", "*", "/", "==", "!=", ">", "<", ">=", "<="];
            if !OPERATORS.contains(&operator.as_str()) {
                return Err(ValidationError(format!("unknown operator '{operator}'")));
            }
            validate_expression(left)?;
            validate_expression(right)?;
        }
        Expression::ToolCall { name, arguments } => {
            check_identifier(name, "tool")?;
            for arg in arguments { validate_expression(arg)?; }
        }
        Expression::FunctionCall { name, arguments } => {
            check_qualified_identifier(name, "function")?;
            for arg in arguments { validate_expression(arg)?; }
        }
        Expression::ListExpression(items) => {
            for item in items { validate_expression(item)?; }
        }
        Expression::MapExpression(entries) => {
            for (k, v) in entries {
                validate_expression(k)?;
                validate_expression(v)?;
            }
        }
        Expression::Variable(name) => check_identifier(name, "variable")?,
        Expression::Recall(key) => check_identifier(key, "memory")?,
        _ => {}
    }
    Ok(())
}

fn validate_ui_tree(root: &UiNode) -> Result<(), ValidationError> {
    let mut ids = HashSet::new();
    let mut stack = vec![root];
    while let Some(node) = stack.pop() {
        if node.node_id < 1 || !ids.insert(node.node_id) {
            return Err(ValidationError(format!("invalid or duplicate UI node id '{}'", node.node_id)));
        }
        let contract = ui_registry::component(node.component_id)
            .ok_or_else(|| ValidationError(format!("unknown UI component '{}'", node.component_id)))?;
        let mut prop_ids = HashSet::new();
        for prop in &node.properties {
            if !prop_ids.insert(prop.property_id) {
                return Err(ValidationError(format!("duplicate UI property on node '{}'", node.node_id)));
            }
            let expected = contract.properties.iter().find(|p| p.id == prop.property_id)
                .ok_or_else(|| ValidationError(format!("property '{}' is not valid for component '{}'", prop.property_id, node.component_id)))?;
            match &prop.value {
                Expression::Literal(val) => {
                    let actual = match val {
                        Value::String(_) => "string",
                        Value::Int(_) => "int",
                        Value::Bool(_) => "bool",
                        _ => return Err(ValidationError("UI property has unsupported literal type".into())),
                    };
                    if actual != expected.type_name {
                        return Err(ValidationError(format!("property '{}' requires {}, got {}", prop.property_id, expected.type_name, actual)));
                    }
                }
                Expression::Variable(_) | Expression::ToolCall { .. } | Expression::FunctionCall { .. } => {
                    // Dynamic properties are allowed — resolved at runtime
                }
                _ => {
                    return Err(ValidationError("UI property has unsupported expression type".into()));
                }
            }
        }
        let mut event_ids = HashSet::new();
        for event in &node.events {
            if !contract.events.contains(&event.event_id) || event.action_id < 1 || !event_ids.insert(event.event_id) {
                return Err(ValidationError(format!("invalid event '{}' on node '{}'", event.event_id, node.node_id)));
            }
        }
        if !node.children.is_empty() && !contract.children {
            return Err(ValidationError(format!("component '{}' cannot have children", node.component_id)));
        }
        stack.extend(node.children.iter());
    }
    Ok(())
}

fn validate_call_graph(program: &Program, _runnables: &HashMap<String, usize>) -> Result<(), ValidationError> {
    let mut graph: HashMap<String, HashSet<String>> = HashMap::new();
    for instruction in &program.instructions {
        match instruction {
            Instruction::Agent(agent) => { graph.insert(agent.name.clone(), runs(&agent.body)); }
            Instruction::Workflow(wf) => { graph.insert(wf.name.clone(), runs(&wf.body)); }
            _ => continue,
        };
    }
    // DFS cycle detection
    let mut state: HashMap<String, u8> = graph.keys().map(|k| (k.clone(), 0u8)).collect();
    for root in graph.keys() {
        if state[root] != 0 { continue; }
        let mut dfs_stack: Vec<(String, Vec<String>)> = vec![(root.clone(), graph[root].iter().cloned().collect())];
        state.insert(root.clone(), 1);
        while let Some((name, deps)) = dfs_stack.last_mut() {
            if deps.is_empty() {
                state.insert(name.clone(), 2);
                dfs_stack.pop();
                continue;
            }
            let dep = deps.pop().unwrap();
            if state.get(&dep) == Some(&1) {
                return Err(ValidationError(format!("workflow cycle detected at '{dep}'")));
            }
            if state.get(&dep) == Some(&0) {
                if dfs_stack.len() >= MAX_CALL_DEPTH {
                    return Err(ValidationError(format!("workflow call depth exceeds {MAX_CALL_DEPTH}")));
                }
                state.insert(dep.clone(), 1);
                let next_deps = graph.get(&dep).cloned().unwrap_or_default().into_iter().collect();
                dfs_stack.push((dep, next_deps));
            }
        }
    }
    Ok(())
}

fn runs(body: &[Instruction]) -> HashSet<String> {
    let mut result = HashSet::new();
    for instruction in body {
        match instruction {
            Instruction::Run(name) => { result.insert(name.clone()); }
            Instruction::If(if_inst) => {
                result.extend(runs(&if_inst.body));
                result.extend(runs(&if_inst.else_body));
            }
            Instruction::While(w) => { result.extend(runs(&w.body)); }
            _ => {}
        }
    }
    result
}
